### PR TITLE
Move CHOLMOD.common_struct from Vector{UInt8} to an actual struct

### PR DESCRIPTION
**Pull request by ettersi**
Related: https://github.com/JuliaLang/LinearAlgebra.jl/issues/649

### CODE REVIEW

**Comment by musm** [comment]
@nanosoldier runbenchmarks(ALL, vs=":master")

**Comment by musm** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/38919#issuecomment-748274237) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/0206113fcadcf9dcb004acf976e569d998ad0637/benchmark/by_hash/f9488fe_vs_114c784/report.md). cc @christopher-dG

File: deps/SuiteSparse_wrapper.c [hunk 1]
```diff
@@ -7,10 +7,6 @@
 #include <string.h>
 #include <cholmod.h>
```
Comment (musm) [inline deps/SuiteSparse_wrapper.c:8]: Unrelated to this PR: can't we just get rid of this file?

**Comment by andreasnoack** [comment]
While there is no doubt that it is nicer to work with a Julia struct like this, the potential issue here is that the common struct changes over time, see https://github.com/DrTimothyAldenDavis/SuiteSparse/blame/master/CHOLMOD/Include/cholmod_core.h. With a shim, we don't have to do anything but compile the shim for each SuiteSparse release but if we layout the struct manually in Julia then we'd have to keep it in sync with the C header file. I'm quite confident that we'll fail to do that.

**Comment by ettersi** [comment]
I guess we could write a test which checks that the C and Julia structs are indeed equivalent. That's probably a good idea anyway, because it is quite easy to make a mistake when translating all the many many fields of `cholmod_common`. 

Is it possible to get Make to compile a library which is used only for testing? Moving `jl_cholmod_common_offset` into such a library and expanding it to test all fields would probably be the cleanest way to implement this.

**Comment by ViralBShah** [comment]
Yes, we can add a check at build time. 

I guess we could also use Clang.jl to auto-generate the struct definition. Perhaps best done whenever we are able to move the sparse stuff into a separate repo.

**Comment by ettersi** [comment]
I'm currently busy teaching, but I'd be happy to pick this up again once the semester is over (around mid-April) if there is any chance that this PR could be accepted. If not (Viral's comment sounds like you might want to wait for a more major overhaul), then I'd greatly appreciate if you let me know.

**Comment by ViralBShah** [comment]
@andreasnoack seems to be not in favour of this for maintainability reasons, which I agree with as well. So the only way to address that is auto-generation with Clang.jl or something - in which case it would have to wait for the separation of repos.

**Comment by ettersi** [comment]
All right, thanks for letting me know!

One minor remark: I interpreted [@andreasnoack 's remark](https://github.com/JuliaLang/julia/issues/32760#issuecomment-743010390) over in Issue #32760 as encouragement for this PR, so I can't deny that I am a little disappointed that apparently the proposed improvement was actually doomed from the beginning. I'm sure this was just miscommunication and quite possibly I am the one to blame here, but it takes two to miscommunicate so I thought I'd share my feelings in case it could be helpful in the future.

**Comment by ViralBShah** [comment]
Yes, that is an argument for having this change. We really do want to get this PR in - but just with a guard against changes to the struct. BTW, building a separate test library for this as you suggest ought to be something that will be acceptable here, since that will provide a guard against struct changes.

File: deps/SuiteSparse_wrapper.c (hunk 1 above)
Comment (ViralBShah) [reply deps/SuiteSparse_wrapper.c:8]: I think we can - but that can be a separate PR.

**Comment by ViralBShah** [comment]
I'm changing my mind on this. This will let us get rid of the SuiteSparse wrapper - gets us really close. @ettersi Can you help us get this over the finish line? It would be good to add some comments in the code that point to where the struct is (perhaps a github link to Tim Davis' code?)

Can we add some checks to the code so that we can do a CHOLMOD version check so that if SuiteSparse is bumped without checking the code, a warning is given - so that whoever bumps will take a look at the warning? I think that would be sufficient to get this done.

**Comment by ettersi** [comment]
I'd love to help, but I won't be able to do so before May. Any chance you guys can wait until then?

**Comment by ViralBShah** [comment]
Yes, there's no urgency. We have a working system right now. So we just get it done when its ready.

**Comment by ettersi** [comment]
Okay, I believe I addressed your concerns. Specifically, I added tests to `CHOLMOD.__init__()` to check that the size and offsets of the C `cholmod_common` and Julia `Common` structs match. I also added some comments to point out the dependencies between the various pieces of code. Let me know if you have any further concerns.

File: stdlib/SuiteSparse/src/cholmod.jl [hunk 2]
```diff
@@ -261,6 +269,33 @@ function __init__()
...
+                 Julia was compiled with a CHOLMOD version whose binary
+                 representation of the cholmod_common_struct is
+                 incompatible with the one currently linked. This might
+                 cause Julia to terminate when working with sparse matrix
+                 factorizations, e.g. solving systems of equations with \\.
+
+                 This problem can be fixed by modifying the Julia
+                 representation of cholmod_common in cholmod.jl, or by
+                 downloading the generic binaries from www.julialang.org,
+                 which ship with the correct versions of all dependencies,
```
Comment (musm) [inline stdlib/SuiteSparse/src/cholmod.jl:295]:
```suggestion
                 which ship with the correct versions of all dependencies.
```

File: stdlib/SuiteSparse/src/cholmod.jl [hunk 3]
```diff
@@ -261,6 +269,33 @@ function __init__()
                  """
         end
 
+        ### Check that the Common and Method Julia structs are consistent with their C counterparts
+        method_offsets = Vector{Csize_t}(undef,16)
```
Comment (musm) [inline stdlib/SuiteSparse/src/cholmod.jl]: Is there a reason this is length 16 and not 15 ? Same with the `common_offsets` array which should store 107 elements I think?

Comment (ettersi) [reply stdlib/SuiteSparse/src/cholmod.jl]: Yes, that's wrong. Fixed it, sorry about that.

**Comment by ViralBShah** [comment]
Is it possible to get rid of the C wrapper code completely?

**Comment by ViralBShah** [comment]
We should also hard-code the version of cholmod for which this struct is coded up, so that if cholmod is bumped in the distribution, one has to bump the version required in cholmod.jl as well - to avoid accidental breakage when the struct goes out of sync when upstream libraries are bumped.

**Comment by ettersi** [comment]
I have added tests in `CHOLMOD.__init__()` which check that the Julia and C struct definitions are compatible (have same size and offsets). That should address the issue of accidental breaks, no?

**Comment by ettersi** [comment]
> Is it possible to get rid of the C wrapper code completely?

Don't think so. I don't see how we could replicate any of the functions in 
https://github.com/JuliaLang/julia/blob/cebee5ab93410f0b1784dfba7d35486a78659f86/deps/SuiteSparse_wrapper.c
in Julia.

**Comment by ViralBShah** [comment]
If the sizes and offsets are the same in the C and Julia codes, can we not just get rid of the wrapper?

**Comment by ettersi** [comment]
No, because we need the wrapper to check that the offsets are indeed the same (unless there's some trick to get around this that I am not aware of?). Also, SuiteSparse defines a `SuiteSparse_long` type that we need to extract.

**Comment by ViralBShah** [comment]
I am imagining that we could write a little C program that calculates all these things and writes them out into a `.jl` file that can be part of our Julia wrappers for SuiteSparse.

An attempt to do was made by @simonbyrne in https://github.com/JuliaPackaging/Yggdrasil/pull/505

**Comment by ViralBShah** [comment]
This PR will also fix #20985

**Comment by ettersi** [comment]
> I am imagining that we could write a little C program that calculates all these things and writes them out into a .jl file that can be part of our Julia wrappers for SuiteSparse.

I guess in this case we might as well auto-generate the struct definitions using Clang.jl. I'd be happy to have a go at this, but if I understand correctly then simonbyrne and Gnimuc already tried and got stuck so I doubt I'd get very far without some very detailed explanations.

**Comment by ettersi** [comment]
Maybe we can just finish off this PR and leave eliminating `SuiteSparse_wrappers.c` to a future PR?

**Comment by musm** [comment]
It does seem that the two issues are orthogonal. @ViralBShah is there a particular reason you would like to hold off until both are updated or was that more of a suggestion for future work? I imagine it would be beneficial to merge this PR as is, and then @ettersi can have the opportunity to tackle eliminating the wrappers in a different PR.

**Comment by musm** [comment]
@ettersi There are some real test  failures that need to be resolved before we can merge this.

**Comment by ViralBShah** [comment]
@musm Agree, nothing I said about replacing the C wrappers should block this. We should merge as soon as it is all green.

**Comment by simonbyrne** [comment]
> I guess in this case we might as well auto-generate the struct definitions using Clang.jl. I'd be happy to have a go at this, but if I understand correctly then simonbyrne and Gnimuc already tried and got stuck so I doubt I'd get very far without some very detailed explanations.

My PR (https://github.com/JuliaPackaging/Yggdrasil/pull/505) attempted to compile and run a C program that would output the offsets: unfortunately that doesn't work because you can't actually run platform-specific code in the BinaryBuilder sandbox.

@Gnimuc's idea to use Clang.jl (https://github.com/JuliaPackaging/Yggdrasil/pull/505#issuecomment-588066990) looks like it works, but would require someone to integrate it with BinaryBuilder.jl

**Comment by Gnimuc** [comment]
I believe it's possible to do cross-compilation with pure Clang.jl, build flags(include dirs, defines, targets) and corresponding artifacts(system headers).

**Comment by ettersi** [comment]
I believe the failed tests are because the `SuiteSparse_wrapper.c` file in the Julia repo is now out of sync with the one in Yggdrasil.
This hypothesis is based on the observation that the first error message in the failed tests is
```
��� Error: Error during initialization of module CHOLMOD
���   exception =
���    could not load symbol "jl_cholmod_method_offsets":
���    /buildworker/worker/tester_linux64/build/bin/../lib/julia/libsuitesparse_wrapper.so: undefined symbol: jl_cholmod_method_offsets
```
I know only little about how Yggdrasil works and how it interacts with base Julia, but Yggdrasil being out of sync with Julia is the only reason I can think of why some tests run into the above while others don't. 

What is the procedure for resolving this issue? Should I create another PR in Yggdrasil with the required changes, and then whoever merges has to make sure they merge both at the same time? Or is there a safer way to do this?

-----

Also, I've moved the offset checks from `CHOLMOD.__init__()` to the cholmod test file. Placing these checks in `__init__()` was the result of some misconceptions I had about how linking works, and I now believe that it makes no sense to run these checks every time someone loads the CHOLMOD package. 

On a related topic, it seems that CHOLMOD occasionally rearranges the order of the `cholmod_common` fields without bumping the major version number (at least that's the impression I got looking at the [blame history](https://github.com/DrTimothyAldenDavis/SuiteSparse/blame/master/CHOLMOD/Include/cholmod_core.h)). So maybe this would require us to be a bit more strict when comparing version numbers [on this line](https://github.com/JuliaLang/julia/blob/4612f87581a0c9f58b7374f32c183a33d0545204/stdlib/SuiteSparse/src/cholmod.jl#L236)?

Long story short, it would probably be good if someone with a better understanding of versioning and linking would have another look at this to make sure I did not mess things up.

**Comment by Gnimuc** [comment]
> @Gnimuc's idea to use Clang.jl ([JuliaPackaging/Yggdrasil#505 (comment)](https://github.com/JuliaPackaging/Yggdrasil/pull/505#issuecomment-588066990)) looks like it works, but would require someone to integrate it with BinaryBuilder.jl ([JuliaPackaging/BinaryBuilder.jl#673](https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/673))

@simonbyrne I just wrote a concrete example using [Clang.jl's new generator](https://github.com/JuliaInterop/Clang.jl/pull/278), indeed, with [some hacks](https://github.com/Gnimuc/SuiteSparseGenerator/blob/4ab1e53f78cf3701e8b9bac4ad0ff9942e0a5806/generator.jl#L71-L84): https://github.com/Gnimuc/SuiteSparseGenerator. 

I now get stuck with the `sysroot` flag, even I passed it to libclang via `--sysroot=...`, it just failed to recognize it.

**Comment by ViralBShah** [comment]
@ettersi The new `SuiteSparse_wrapper.c` for now should just be a PR to the Yggdrasil repo - specifically, replace this file with the new one: https://github.com/JuliaPackaging/Yggdrasil/tree/master/S/SuiteSparse/SuiteSparse/bundled/SuiteSparse_wrapper

The PR will kick off the build process. Separately, we should tighten the version checking to closely follow CHOLMOD down to the patch level for now. Both the Yggdrasil PRs and this one should be merged in close succession - with Yggdrasil first. I'm happy to help with the merging.

**Comment by ettersi** [comment]
All right, I tightened the version number check and created the Yggdrasil pull request (https://github.com/JuliaPackaging/Yggdrasil/pull/2848). I believe this should be good to go now?

**Comment by ViralBShah** [comment]
I've merged the Yggdrasil PR, which should propagate into the relevant places in the next hour or so. We will need all green here based on that PR and then merge this one.

**Comment by ViralBShah** [comment]
I merged https://github.com/JuliaRegistries/General/pull/34441 by hand.

**Comment by ViralBShah** [comment]
This PR needs to update https://github.com/JuliaLang/julia/blob/master/stdlib/SuiteSparse_jll/Project.toml. The version needs to be changed to `5.8.1+2` so that it picks up the new SuiteSparse binaries.

**Comment by ettersi** [comment]
Last remaining errors are:
```
make: *** No rule to make target 'win-extras'.  Stop.
```
Unrelated? :crossed_fingers:

**Comment by ViralBShah** [comment]
Maybe rebase to the latest master and hope that goes away?

**Comment by ettersi** [comment]
Did the rebase. The tests which are failing now were added in https://github.com/JuliaLang/julia/pull/39301, and it seems there are some issues with these tests, see the recent discussion in https://github.com/JuliaLang/julia/pull/40504. I guess we wait until these issues are resolved and try again.

**Comment by ettersi** [comment]
All green! :tada::tada::tada:

**Comment by ettersi** [comment]
Awesome, thanks everyone for all your help! As you could probably tell, this was my first (significant) contribution to Julia, and thanks to all of you this has been a very enjoyable and rewarding experience!

**Comment by ViralBShah** [comment]
Thanks @ettersi for the contribution and staying with the process - which can be a bit long at times. But as a result now we have one more person who understands our sparse solver infrastructure :-)
